## Tarea 1_ Red neuronal Perceptrón multicapa con Keras


**Se importan librerias necesarias para el funcionamiento de nuestro MLP**

In [ ]:
%tensorflow_version 2.x

Colab only includes TensorFlow 2.x; %tensorflow_version has no effect.


In [ ]:
import tensorflow as tf
import numpy as np
import matplotlib.pyplot as plt
%matplotlib inline
from sklearn.preprocessing import LabelEncoder
from sklearn.preprocessing import OneHotEncoder

**Se cargan desde los datasets de keras y tensorflow el conjunto de datos fashion_mnist**

In [ ]:
from keras.datasets import fashion_mnist
(x_train, y_train), (x_test, y_test) = fashion_mnist.load_data()

4431872/4422102 [==============================] - 0s 0us/step


**El preprocesamiento de datos se puede hacer más simple utilizando Keras utilizando to_categorical, un método de Keras que nos ayudará a convertir nuestras etiquetas, no sin antes, normalizar nuestros datos entre 0 y 1 (en los datos en X)**

In [ ]:
from tensorflow.keras.utils import to_categorical

 
x_train = x_train.reshape((60000, 784))
x_train = x_train.astype('float32') / 255
y_train = to_categorical(y_train)
 
x_test = x_test.reshape((10000, 784))
x_test = x_test.astype('float32') / 255
y_test = to_categorical(y_test)

**Mezclamos nuestros conjuntos y armamos nuestro conjunto de validación**

In [ ]:
import numpy as np
 
np.random.seed(42)
 
# barajamos el dataset
permutation = np.random.permutation(x_train.shape[0])
x_train = x_train[permutation]
y_train = y_train[permutation]
 
# obtenemos el conjunto de validación
num_val = 10000
x_val = x_train[:num_val]
x_train = x_train[num_val:]
y_val = y_train[:num_val]
y_train = y_train[num_val:]

# Arquitectura

** A diferencia de ternsorflow, podemos implementar un MLP mucho más fácil únicamente agregando capas densas en un modelo secuencial**

In [ ]:
from keras import models, layers
 
model = models.Sequential()
model.add(layers.Dense(340, activation='relu', input_shape=(784,))) # 28x28=784
model.add(layers.Dense(200, activation='relu'))
model.add(layers.Dense(100, activation='relu'))
model.add(layers.Dense(10, activation='softmax'))

**Se compila y se entrena el modelo utilizando el optimizador de Adam**

In [ ]:
model.compile(optimizer = 'Adam',
             loss = 'categorical_crossentropy',
             metrics = ['accuracy'])

In [ ]:
train_log = model.fit(x_train, y_train, 
                      epochs=250, batch_size=228,
                     validation_data=(x_val, y_val))

Epoch 1/250
220/220 [==============================] - 6s 9ms/step - loss: 0.5576 - accuracy: 0.8080 - val_loss: 0.4262 - val_accuracy: 0.8505
Epoch 2/250
220/220 [==============================] - 1s 5ms/step - loss: 0.3858 - accuracy: 0.8624 - val_loss: 0.3747 - val_accuracy: 0.8634
Epoch 3/250
220/220 [==============================] - 1s 4ms/step - loss: 0.3396 - accuracy: 0.8760 - val_loss: 0.3616 - val_accuracy: 0.8653
Epoch 4/250
220/220 [==============================] - 1s 4ms/step - loss: 0.3073 - accuracy: 0.8873 - val_loss: 0.3192 - val_accuracy: 0.8832
Epoch 5/250
220/220 [==============================] - 1s 4ms/step - loss: 0.2872 - accuracy: 0.8937 - val_loss: 0.3330 - val_accuracy: 0.8792
Epoch 6/250
220/220 [==============================] - 1s 4ms/step - loss: 0.2699 - accuracy: 0.8994 - val_loss: 0.3355 - val_accuracy: 0.8805
Epoch 7/250
220/220 [==============================] - 1s 4ms/step - loss: 0.2595 - accuracy: 0.9033 - val_loss: 0.3186 - val_accuracy: 0.8862

# Conclusiones

**Entrenar un modelo utilizando Keras es mucho más sencillo y el resultado tiende a veces a ser mejor, sin embargo, es muy notorio como hacer uso de métodos ya incorporados en keras, en muchas ocasiones tienda a opacar el proceso que estás realizando, es decir, si no conoces el funcionamiento real del método, quedas atado a modificar unos cuantos parámetros sin posibilidad de ver qué es lo que realmente ocurre. Para redes más complejas puede ser necesario hacer uso de tensorflow puro para poder modificar hiperparámetros y diseñar arquitecturas más complejas, sin embargo, el resultado en este MLP no varió mucho de su versión puramente en TF2. Lo interesante de este modelo de Keras fue que en algunas épocas la exactitud llegó a estar incluso en 100% (de la 186 a la 220), lo cual llega a ser extraño.**